File: platform/packages_mjs/fetch_undici_gemini_openai_protocols/src/response.ts
/**
 * Response Schema Mapping
 *
 * Converts between OpenAI chat completion responses and Gemini generateContent responses.
 * Handles structured output, finish reasons, and usage statistics.
 */

import { logger } from 'fetch-undici'
import type {
  OpenAIChatResponse,
  OpenAIChoice,
  OpenAIUsage,
  OpenAIMessage,
  GeminiGenerateResponse,
  GeminiCandidate,
  GeminiUsageMetadata,
  GeminiContent,
  GeminiFunctionCallPart,
  TranslationResult,
} from './types.js'
import {
  geminiFunctionCallsToOpenAI,
  isGeminiFunctionCallPart,
  generateToolCallId,
} from './tools.js'
import { mapGeminiRoleToOpenAI } from './messages.js'

const log = logger.create('fetch-undici-gemini-openai-protocols:response', import.meta.url)

// =============================================================================
// Finish Reason Mapping
// =============================================================================

type OpenAIFinishReason = 'stop' | 'length' | 'tool_calls' | 'content_filter' | 'function_call' | null
type GeminiFinishReason = 'STOP' | 'MAX_TOKENS' | 'SAFETY' | 'RECITATION' | 'OTHER' | undefined

/**
 * Map Gemini finish reason to OpenAI finish reason
 */
export function mapGeminiFinishReason(reason: GeminiFinishReason): OpenAIFinishReason {
  switch (reason) {
    case 'STOP':
      return 'stop'
    case 'MAX_TOKENS':
      return 'length'
    case 'SAFETY':
    case 'RECITATION':
      return 'content_filter'
    case 'OTHER':
    default:
      return null
  }
}

/**
 * Map OpenAI finish reason to Gemini finish reason
 */
export function mapOpenAIFinishReason(reason: OpenAIFinishReason): GeminiFinishReason {
  switch (reason) {
    case 'stop':
      return 'STOP'
    case 'length':
      return 'MAX_TOKENS'
    case 'content_filter':
      return 'SAFETY'
    case 'tool_calls':
    case 'function_call':
      return 'STOP' // Tool calls are considered successful completion
    default:
      return 'OTHER'
  }
}

// =============================================================================
// Usage Statistics Mapping
// =============================================================================

/**
 * Convert Gemini usage metadata to OpenAI usage
 */
export function geminiUsageToOpenAI(metadata: GeminiUsageMetadata): OpenAIUsage {
  return {
    prompt_tokens: metadata.promptTokenCount,
    completion_tokens: metadata.candidatesTokenCount,
    total_tokens: metadata.totalTokenCount,
  }
}

/**
 * Convert OpenAI usage to Gemini usage metadata
 */
export function openAIUsageToGemini(usage: OpenAIUsage): GeminiUsageMetadata {
  return {
    promptTokenCount: usage.prompt_tokens,
    candidatesTokenCount: usage.completion_tokens,
    totalTokenCount: usage.total_tokens,
  }
}

// =============================================================================
// Gemini → OpenAI Response Translation
// =============================================================================

/**
 * Extract text content from Gemini content
 */
function extractTextContent(content: GeminiContent): string {
  return content.parts
    .filter((p): p is { text: string } => 'text' in p)
    .map(p => p.text)
    .join('')
}

/**
 * Extract function call parts from Gemini content
 */
function extractFunctionCallParts(content: GeminiContent): GeminiFunctionCallPart[] {
  return content.parts.filter(isGeminiFunctionCallPart)
}

/**
 * Convert Gemini candidate to OpenAI choice
 */
function geminiCandidateToOpenAIChoice(
  candidate: GeminiCandidate,
  index: number,
  idGenerator: () => string = generateToolCallId
): OpenAIChoice {
  const { content, finishReason } = candidate

  // Build the message
  const message: OpenAIMessage = {
    role: mapGeminiRoleToOpenAI(content.role),
    content: extractTextContent(content) || null,
  }

  // Check for function calls
  const functionCallParts = extractFunctionCallParts(content)
  if (functionCallParts.length > 0) {
    const { data: toolCalls } = geminiFunctionCallsToOpenAI(functionCallParts, idGenerator)
    message.tool_calls = toolCalls
    // If there are tool calls, the finish reason should reflect that
    return {
      index,
      message,
      finish_reason: toolCalls.length > 0 ? 'tool_calls' : mapGeminiFinishReason(finishReason),
    }
  }

  return {
    index,
    message,
    finish_reason: mapGeminiFinishReason(finishReason),
  }
}

/**
 * Convert Gemini generateContent response to OpenAI chat completion response
 */
export function geminiToOpenAIResponse(
  response: GeminiGenerateResponse,
  options: {
    model?: string
    idGenerator?: () => string
  } = {}
): TranslationResult<OpenAIChatResponse> {
  log.debug('Converting Gemini response to OpenAI', {
    candidateCount: response.candidates?.length ?? 0,
  })

  const warnings: string[] = []
  const { model = 'gemini-model', idGenerator = generateToolCallId } = options

  // Generate a unique ID
  const id = `chatcmpl-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`

  // Convert candidates to choices
  const choices: OpenAIChoice[] = (response.candidates || []).map((candidate, index) =>
    geminiCandidateToOpenAIChoice(candidate, index, idGenerator)
  )

  // Handle blocked responses
  if (response.promptFeedback?.blockReason) {
    warnings.push(`Prompt blocked: ${response.promptFeedback.blockReason}`)
  }

  const result: OpenAIChatResponse = {
    id,
    object: 'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model,
    choices,
  }

  // Add usage if available
  if (response.usageMetadata) {
    result.usage = geminiUsageToOpenAI(response.usageMetadata)
  }

  log.debug('Converted to OpenAI response', {
    id: result.id,
    choiceCount: choices.length,
    hasUsage: !!result.usage,
  })

  return {
    data: result,
    warnings,
  }
}

// =============================================================================
// OpenAI → Gemini Response Translation
// =============================================================================

/**
 * Convert OpenAI message to Gemini content
 */
function openAIMessageToGeminiContent(message: OpenAIMessage): GeminiContent {
  const parts: GeminiContent['parts'] = []

  // Add text content
  if (message.content) {
    parts.push({ text: message.content })
  }

  // Add function calls
  if (message.tool_calls) {
    for (const toolCall of message.tool_calls) {
      if (toolCall.type === 'function') {
        let args: Record<string, unknown>
        try {
          args = JSON.parse(toolCall.function.arguments)
        } catch {
          args = {}
        }
        parts.push({
          functionCall: {
            name: toolCall.function.name,
            args,
          },
        })
      }
    }
  }

  // Handle legacy function_call
  if (message.function_call) {
    let args: Record<string, unknown>
    try {
      args = JSON.parse(message.function_call.arguments)
    } catch {
      args = {}
    }
    parts.push({
      functionCall: {
        name: message.function_call.name,
        args,
      },
    })
  }

  return {
    role: message.role === 'assistant' ? 'model' : 'user',
    parts,
  }
}

/**
 * Convert OpenAI choice to Gemini candidate
 */
function openAIChoiceToGeminiCandidate(choice: OpenAIChoice): GeminiCandidate {
  return {
    content: openAIMessageToGeminiContent(choice.message),
    finishReason: mapOpenAIFinishReason(choice.finish_reason),
    index: choice.index,
  }
}

/**
 * Convert OpenAI chat completion response to Gemini generateContent response
 */
export function openAIToGeminiResponse(
  response: OpenAIChatResponse
): TranslationResult<GeminiGenerateResponse> {
  log.debug('Converting OpenAI response to Gemini', {
    id: response.id,
    choiceCount: response.choices.length,
  })

  const warnings: string[] = []

  const result: GeminiGenerateResponse = {
    candidates: response.choices.map(openAIChoiceToGeminiCandidate),
  }

  // Add usage metadata if available
  if (response.usage) {
    result.usageMetadata = openAIUsageToGemini(response.usage)
  }

  log.debug('Converted to Gemini response', {
    candidateCount: result.candidates.length,
    hasUsage: !!result.usageMetadata,
  })

  return {
    data: result,
    warnings,
  }
}

// =============================================================================
// Structured Output Helpers
// =============================================================================

/**
 * Extract JSON from response content (handles markdown code blocks)
 */
export function extractJSON<T = unknown>(content: string): T | null {
  // Try direct parse first
  try {
    return JSON.parse(content) as T
  } catch {
    // Try to extract from markdown code block
    const codeBlockMatch = content.match(/```(?:json)?\s*([\s\S]*?)```/)
    if (codeBlockMatch) {
      try {
        return JSON.parse(codeBlockMatch[1].trim()) as T
      } catch {
        // Fall through
      }
    }

    // Try to find JSON object in content
    const jsonMatch = content.match(/\{[\s\S]*\}/)
    if (jsonMatch) {
      try {
        return JSON.parse(jsonMatch[0]) as T
      } catch {
        // Fall through
      }
    }

    return null
  }
}

/**
 * Validate response against JSON schema (basic validation)
 */
export function validateAgainstSchema(
  data: unknown,
  _schema: Record<string, unknown>
): { valid: boolean; errors: string[] } {
  // Basic type validation - full JSON Schema validation would require a library
  if (data === null || data === undefined) {
    return { valid: false, errors: ['Data is null or undefined'] }
  }

  // For now, just check it's an object (full validation would use ajv or similar)
  if (typeof data !== 'object') {
    return { valid: false, errors: ['Data is not an object'] }
  }

  return { valid: true, errors: [] }
}

File: platform/packages_py/fetch_httpx_gemini_openai_protocols/fetch_httpx_gemini_openai_protocols/response.py
"""
Response Schema Mapping

Converts between OpenAI chat completion responses and Gemini generateContent responses.
Handles structured output, finish reasons, and usage statistics.
"""

from __future__ import annotations

import json
import random
import re
import time
from collections.abc import Callable
from typing import Any

from fetch_httpx import create_logger

from .messages import map_gemini_role_to_openai
from .tools import (
    gemini_function_calls_to_openai,
    generate_tool_call_id,
    is_gemini_function_call_part,
)
from .types import (
    GeminiCandidate,
    GeminiContent,
    GeminiFunctionCallPart,
    GeminiGenerateResponse,
    GeminiUsageMetadata,
    OpenAIChatResponse,
    OpenAIChoice,
    OpenAIMessage,
    OpenAIUsage,
    TranslationResult,
)

logger = create_logger("fetch_httpx_gemini_openai_protocols.response", __file__)

# =============================================================================
# Finish Reason Mapping
# =============================================================================

OpenAIFinishReason = str | None
GeminiFinishReason = str | None


def map_gemini_finish_reason(reason: GeminiFinishReason) -> OpenAIFinishReason:
    """Map Gemini finish reason to OpenAI finish reason."""
    mapping = {
        "STOP": "stop",
        "MAX_TOKENS": "length",
        "SAFETY": "content_filter",
        "RECITATION": "content_filter",
        "OTHER": None,
    }
    return mapping.get(reason) if reason else None


def map_openai_finish_reason(reason: OpenAIFinishReason) -> GeminiFinishReason:
    """Map OpenAI finish reason to Gemini finish reason."""
    mapping = {
        "stop": "STOP",
        "length": "MAX_TOKENS",
        "content_filter": "SAFETY",
        "tool_calls": "STOP",
        "function_call": "STOP",
    }
    return mapping.get(reason, "OTHER") if reason else None


# =============================================================================
# Usage Statistics Mapping
# =============================================================================


def gemini_usage_to_openai(metadata: GeminiUsageMetadata) -> OpenAIUsage:
    """Convert Gemini usage metadata to OpenAI usage."""
    return {
        "prompt_tokens": metadata.get("promptTokenCount", 0),
        "completion_tokens": metadata.get("candidatesTokenCount", 0),
        "total_tokens": metadata.get("totalTokenCount", 0),
    }


def openai_usage_to_gemini(usage: OpenAIUsage) -> GeminiUsageMetadata:
    """Convert OpenAI usage to Gemini usage metadata."""
    return {
        "promptTokenCount": usage.get("prompt_tokens", 0),
        "candidatesTokenCount": usage.get("completion_tokens", 0),
        "totalTokenCount": usage.get("total_tokens", 0),
    }


# =============================================================================
# Gemini → OpenAI Response Translation
# =============================================================================


def _extract_text_content(content: GeminiContent) -> str:
    """Extract text content from Gemini content."""
    parts = content.get("parts", [])
    return "".join(
        part.get("text", "")
        for part in parts
        if "text" in part
    )


def _extract_function_call_parts(content: GeminiContent) -> list[GeminiFunctionCallPart]:
    """Extract function call parts from Gemini content."""
    parts = content.get("parts", [])
    return [part for part in parts if is_gemini_function_call_part(part)]


def _gemini_candidate_to_openai_choice(
    candidate: GeminiCandidate,
    index: int,
    id_generator: Callable[[], str] | None = None,
) -> OpenAIChoice:
    """Convert Gemini candidate to OpenAI choice."""
    if id_generator is None:
        id_generator = generate_tool_call_id

    content = candidate.get("content", {"role": "model", "parts": []})
    finish_reason = candidate.get("finishReason")

    # Build the message
    message: OpenAIMessage = {
        "role": map_gemini_role_to_openai(content.get("role", "model")),
        "content": _extract_text_content(content) or None,
    }

    # Check for function calls
    function_call_parts = _extract_function_call_parts(content)
    if function_call_parts:
        result = gemini_function_calls_to_openai(function_call_parts, id_generator)
        tool_calls = result.data
        message["tool_calls"] = tool_calls
        return {
            "index": index,
            "message": message,
            "finish_reason": "tool_calls" if tool_calls else map_gemini_finish_reason(finish_reason),
        }

    return {
        "index": index,
        "message": message,
        "finish_reason": map_gemini_finish_reason(finish_reason),
    }


def gemini_to_openai_response(
    response: GeminiGenerateResponse,
    model: str = "gemini-model",
    id_generator: Callable[[], str] | None = None,
) -> TranslationResult[OpenAIChatResponse]:
    """Convert Gemini generateContent response to OpenAI chat completion response."""
    logger.debug(
        "Converting Gemini response to OpenAI",
        context={"candidate_count": len(response.get("candidates", []))},
    )

    warnings: list[str] = []

    # Generate a unique ID
    random_suffix = "".join(random.choices("abcdefghijklmnopqrstuvwxyz0123456789", k=7))
    response_id = f"chatcmpl-{int(time.time())}-{random_suffix}"

    # Convert candidates to choices
    candidates = response.get("candidates", [])
    choices: list[OpenAIChoice] = [
        _gemini_candidate_to_openai_choice(candidate, index, id_generator)
        for index, candidate in enumerate(candidates)
    ]

    # Handle blocked responses
    prompt_feedback = response.get("promptFeedback")
    if prompt_feedback and prompt_feedback.get("blockReason"):
        warnings.append(f"Prompt blocked: {prompt_feedback['blockReason']}")

    result: OpenAIChatResponse = {
        "id": response_id,
        "object": "chat.completion",
        "created": int(time.time()),
        "model": model,
        "choices": choices,
    }

    # Add usage if available
    usage_metadata = response.get("usageMetadata")
    if usage_metadata:
        result["usage"] = gemini_usage_to_openai(usage_metadata)

    logger.debug(
        "Converted to OpenAI response",
        context={
            "id": result["id"],
            "choice_count": len(choices),
            "has_usage": "usage" in result,
        },
    )

    return TranslationResult(data=result, warnings=warnings)


# =============================================================================
# OpenAI → Gemini Response Translation
# =============================================================================


def _openai_message_to_gemini_content(message: OpenAIMessage) -> GeminiContent:
    """Convert OpenAI message to Gemini content."""
    parts: list[dict[str, Any]] = []

    # Add text content
    content = message.get("content")
    if content:
        parts.append({"text": content})

    # Add function calls
    tool_calls = message.get("tool_calls")
    if tool_calls:
        for tool_call in tool_calls:
            if tool_call.get("type") == "function":
                func = tool_call.get("function", {})
                try:
                    args = json.loads(func.get("arguments", "{}"))
                except json.JSONDecodeError:
                    args = {}
                parts.append({
                    "functionCall": {
                        "name": func.get("name", ""),
                        "args": args,
                    }
                })

    # Handle legacy function_call
    function_call = message.get("function_call")
    if function_call:
        try:
            args = json.loads(function_call.get("arguments", "{}"))
        except json.JSONDecodeError:
            args = {}
        parts.append({
            "functionCall": {
                "name": function_call.get("name", ""),
                "args": args,
            }
        })

    role = message.get("role", "assistant")
    return {
        "role": "model" if role == "assistant" else "user",
        "parts": parts,
    }


def _openai_choice_to_gemini_candidate(choice: OpenAIChoice) -> GeminiCandidate:
    """Convert OpenAI choice to Gemini candidate."""
    message = choice.get("message", {})
    return {
        "content": _openai_message_to_gemini_content(message),
        "finishReason": map_openai_finish_reason(choice.get("finish_reason")),
        "index": choice.get("index", 0),
    }


def openai_to_gemini_response(
    response: OpenAIChatResponse,
) -> TranslationResult[GeminiGenerateResponse]:
    """Convert OpenAI chat completion response to Gemini generateContent response."""
    logger.debug(
        "Converting OpenAI response to Gemini",
        context={
            "id": response.get("id"),
            "choice_count": len(response.get("choices", [])),
        },
    )

    warnings: list[str] = []

    choices = response.get("choices", [])
    result: GeminiGenerateResponse = {
        "candidates": [_openai_choice_to_gemini_candidate(choice) for choice in choices],
    }

    # Add usage metadata if available
    usage = response.get("usage")
    if usage:
        result["usageMetadata"] = openai_usage_to_gemini(usage)

    logger.debug(
        "Converted to Gemini response",
        context={
            "candidate_count": len(result["candidates"]),
            "has_usage": "usageMetadata" in result,
        },
    )

    return TranslationResult(data=result, warnings=warnings)


# =============================================================================
# Structured Output Helpers
# =============================================================================


def extract_json(content: str) -> Any | None:
    """Extract JSON from response content (handles markdown code blocks)."""
    # Try direct parse first
    try:
        return json.loads(content)
    except json.JSONDecodeError:
        pass

    # Try to extract from markdown code block
    code_block_match = re.search(r"```(?:json)?\s*([\s\S]*?)```", content)
    if code_block_match:
        try:
            return json.loads(code_block_match.group(1).strip())
        except json.JSONDecodeError:
            pass

    # Try to find JSON object in content
    json_match = re.search(r"\{[\s\S]*\}", content)
    if json_match:
        try:
            return json.loads(json_match.group(0))
        except json.JSONDecodeError:
            pass

    return None


def validate_against_schema(
    data: Any,
    schema: dict[str, Any],
) -> tuple[bool, list[str]]:
    """Validate response against JSON schema (basic validation)."""
    # Basic type validation - full JSON Schema validation would require a library
    if data is None:
        return False, ["Data is null or undefined"]

    # For now, just check it's an object (full validation would use jsonschema)
    if not isinstance(data, dict):
        return False, ["Data is not an object"]

    return True, []
